Implemented CupertinoButton new styles/sizes (fixes #92525)

kerberjg opened the pull request:
This PR fixes #92525 and introduces the following changes according to [latest iOS HIG](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS):

- `CupertinoButton` now has a `size` property (type `enum CupertinoButtonSize`, values sm/md/lg, default `lg`) that allows the devs to apply new iOS 15+ button styles
- Previously `CupertinoButton` had a larger padding when no background color was specified. With the new HIG, that is no longer the case
- `CupertinoButton` now has a `.tinted` constructor that renders a translucent background (transparency % is brightness-dependent) and uses a different foreground color compared to `.filled`
- `CupertinoButton` now uses the `actionTextStyle` TextStyle from the given theme
- `CupertinoButton`'s child IconTheme's size will always be x1.2 the given TextStyle's size
- `CupertinoTextThemeData` now has a `actionSmallTextStyle` property to use with small buttons (including a default `_kDefaultActionSmallTextStyle` TextStyle)


Preview & example:

![image](https://github.com/user-attachments/assets/0985eb19-c091-41f5-bd98-0de196b7e403)

> **NOTE**: there is a discrepancy in dark mode button foreground color between the default CupertinoTheme and the HIG. A separate issue will be opened for this.

~EDIT: issue reported here https://github.com/flutter/flutter/issues/152846~
EDIT2: fixed by #153039 !

![image](https://github.com/user-attachments/assets/d671d7b4-bb2f-4b38-9464-ee1b04927304)

## Example
```dart
import 'package:flutter/cupertino.dart';

const Widget body = Row(
  mainAxisSize: MainAxisSize.min,
  mainAxisAlignment: MainAxisAlignment.center,
  children: <Widget>[
    Icon(
      CupertinoIcons.play_fill,
    ),
    Text("Play"),
  ],
);

void main() =>
  runApp(
    CupertinoApp(
      home: Container(
        child: Wrap(
        direction: Axis.horizontal,
        children: <Widget>[
          // header
          Text(''),
          Text('Plain'),
          Text('Grey'),
          Text('Tinted'),
          Text('Filled'),
          // small
          Text('Small'),
          CupertinoButton(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
            color: CupertinoColors.systemGrey,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
          ),
          CupertinoButton.filled(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.small,
          ),
          // medium
          Text('Medium'),
          CupertinoButton(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
            color: CupertinoColors.systemGrey,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
          ),
          CupertinoButton.filled(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.medium,
          ),
          // large
          Text('Large'),
          CupertinoButton(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.large,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            color: CupertinoColors.systemGrey,
            size: CupertinoButtonSize.large,
          ),
          CupertinoButton.tinted(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.large,
          ),
          CupertinoButton.filled(
            child: body,
            onPressed: () {},
            size: CupertinoButtonSize.large,
          ),
        ].map((Widget w) => SizedBox(width: 110, height: 70, child: Center(child: w))).toList(),
      ),
      )
    ),
  );

```


*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

kerberjg committed:
Implemented CupertinoButton size

CupertinoButton now takes a [size] property that allows it to be sized small/medium/large as per iOS 17 style guidelines

This obsoletes the previous behavior where, if a CupertinoButton had no background, it would have a larger padding.
Now padding is set based on the [size] property as derived from iOS HIG.

kerberjg committed:
Implemented CupertinoButton.tinted constructor

Adds a 'tinted' variant to CupertinoButton as per iOS 17 HIG.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -12,19 +12,71 @@ import 'colors.dart';
+  large,
+}
+
+const Map<CupertinoButtonSize, EdgeInsetsGeometry> _kCupertinoButtonPadding = <CupertinoButtonSize, EdgeInsetsGeometry>{
```
These maps that store raw style values should probably be made public. It makes it easier for developers making custom widgets to use, either by just referencing them in the docs or pulling them in directly in the code.

MitchellGoodwin replied:
We tend to make these kind of values private, and it's usually good practice to do so for writing app code, but we've received feedback that in the framework code it's a tad annoying.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -139,9 +220,14 @@ class CupertinoButton extends StatefulWidget {
   /// The radius of the button's corners when it has a background color.
   ///
-  /// Defaults to round corners of 8 logical pixels.
+  /// Defaults to [_kCupertinoButtonSizeBorderRadius], based on [size].
```
Dart doc doesn't pick up bracket references to private references well. No issue if we make `_kCupertinoButtonSizeBorderRadius` public.

MitchellGoodwin reviewed (commented):
These look great! I did a first pass at the implementation without going too deep into the details, and made some comments.

I like all of the styling values be in maps for the different size styles at the top. It makes it easy to look through and update the values later if HIGs changes. Just throwing an idea out here, but I wonder if instead of splitting up maps by property type, like `_kCupertinoButtonSizeBorderRadius` and `_kCupertinoButtonMinSize`, we instead of have something like Material's `ButtonStyle`, and we have something like a `CupertinoButtonStyle` class that contains all of the properties for each configuration in one place, for the different sizes as well as filled, outlined and so on. It might be a little bit easier to search through or use in other widgets if needed. The button could then have a `_style` property that gets set based on the constructor and properties. For instance for a filled button of medium size, it does something like this
```
_style = CupertinoButtonStyle.filled.mergeWith(CupertinoButtonStyle.medium)
```
Then throughout the code it can just check with the one `_style` collection rather than having to do multiple checks for the button type. 
```
final Color? backgroundColor =  _style.backgroundColor
```

kerberjg committed:
Fixed CupertinoButton brightness detection

kerberjg committed:
Made CupertinoButton default constants public

kerberjg committed:
Fixed CupertinoButton docs

kerberjg committed:
Renamed `CupertinoButton.size` to `sizeStyle`

kerberjg committed:
Updated Cupertino actionSmallTextStyle

kerberjg commented:
@MitchellGoodwin Thanks for the review! I agree it would be good to have a `CupertinoButtonStyle`, also because that's be closer to how iOS' `UIButton` works - I think it'd be worthwhile to open a separate issue/PR to handle that properly, and for now just introduce a minimal HIG-compliant version of the new properties in this one.

I addressed some of your feedback in the latest commits and left you some questions on others

kerberjg committed:
Fixed `CupertinoButton` tests

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -297,8 +394,15 @@ class _CupertinoButtonState extends State<CupertinoButton> with SingleTickerProv
+    ).copyWith(color: foregroundColor);
+    final IconThemeData iconTheme = IconTheme.of(context).copyWith(
+      color: foregroundColor,
+      size: textStyle.fontSize! * 1.2,
```
```suggestion
      size: ( textStyle.fontSize ?? 14 ) * 1.2,
```
14 would probably be an ok default. It *could* scale up or down depending on `sizeStyle` but that might be overengineering for the case where somebody deliberately removed theme values.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -10,20 +10,73 @@ import 'package:flutter/widgets.dart';
+  large,
+}
+
+const Map<CupertinoButtonSize, EdgeInsetsGeometry> kCupertinoButtonPadding = <CupertinoButtonSize, EdgeInsetsGeometry>{
```
These would need some documentation, pointing out they are style defaults at different `[CupertinoButtonSize]` values. Also include that they come from the iOS 17 HIGs.

MitchellGoodwin commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -27,8 +27,8 @@ void main() {
-      // 1 10px character + 16px * 2 is smaller than the default 44px minimum.
-      const Size.square(44.0),
+      // 1 10px character + 20px * 2 = 50.0
+      const Size(50.0, 44.0),
```
Is this going up because of the padding from `kCupertinoButtonPadding` for the large sized style?

MitchellGoodwin reviewed (commented):
Looks like at least one test is failing for the button, and there's some analyzer issues. Take a look at the 'Linux Analyze' check for the analyzer. That check isn't too bad to look through, but let me know if you have issues finding other errors. The checks can be a slog to look through until you are used to it.

MitchellGoodwin commented:
> for now just introduce a minimal HIG-compliant version of the new properties in this one.

Let's do that. On a second look, I think this implementation is more than sufficient.

kerberjg committed:
Added docs to CupertinoButton constants

kerberjg committed:
Set default `IconTheme` size in `CupertinoButton`

kerberjg committed:
Fixed CupertinoButton analyzer issues

victorsanni commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -9,21 +9,85 @@ import 'package:flutter/foundation.dart';
+
+
+/// The relative values needed to transform a color to it's equivalent focus outline color.
+const double kCupertinoFocusColorOpacity = 0.80,
```
These constants are used everywhere for focus in cupertino: switch, checkbox, radio, button, etc. If we are making them public, they should probably be defined in one place (possibly `constants.dart`) and made public there instead.

victorsanni replied:
I'm wondering in general where these public constants should live. Maybe the ones directly relevant to CupertinoButton should live here, and everything else move to cupertino/constants.dart?

victorsanni replied:
Opened https://github.com/flutter/flutter/pull/153115 to kickstart this move.

victorsanni commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -9,21 +9,85 @@ import 'package:flutter/foundation.dart';
+enum CupertinoButtonSize {
+  /// Displays a smaller button with round sides and smaller text (uses [CupertinoTextThemeData.actionSmallTextStyle]).
+  small,
+  /// Displays a medium sized button with round sides and regular-sized text
```
Nit that will inevitably come up at some point: all comments should end in a period.

victorsanni commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -54,20 +118,48 @@ class CupertinoButton extends StatefulWidget {
+    this.sizeStyle = CupertinoButtonSize.large,
+    this.padding,
+    this.color,
+    this.disabledColor = CupertinoColors.quaternarySystemFill,
```
Does this default disabled color also match native iOS in dark mode? If not, consider using CupertinoDynamicColor.resolve to make it resolve differently in light vs. dark modes.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -9,21 +9,85 @@ import 'package:flutter/foundation.dart';
+// Measured against iOS 17 guidelines (https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS)
+
+/// The size of a [CupertinoButton].
+/// Based on the iOS (17) Human Interface Guidelines (https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS)
```
```suggestion
/// Based on the iOS (17) [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS).
```
nit: Prefer formatted links, plus punctuation.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -9,21 +9,85 @@ import 'package:flutter/foundation.dart';
+}
+
+/// The padding values for the different [CupertinoButtonSize]s.
+/// Based on the iOS (17) Human Interface Guidelines (https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS)
```
Same here, and below for the similar constants.

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -9,21 +9,85 @@ import 'package:flutter/foundation.dart';
+             kCupertinoFocusColorBrightness = 0.69,
+             kCupertinoFocusColorSaturation = 0.835;
+
+/// Opacity values for the background of a [CupertinoButton.tinted]
```
```suggestion
/// Opacity values for the background of a [CupertinoButton.tinted].
```
nit

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -139,9 +232,14 @@ class CupertinoButton extends StatefulWidget {
+  /// Defaults to [kCupertinoButtonSizeBorderRadius], based on [sizeStyle].
   final BorderRadius? borderRadius;
 
+  /// The size of the button
+  ///
+  /// Defaults to [CupertinoButtonSize.large]
```
```suggestion
  /// The size of the button.
  ///
  /// Defaults to [CupertinoButtonSize.large].
```
nit

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -317,12 +431,10 @@ class _CupertinoButtonState extends State<CupertinoButton> with SingleTickerProv
-                    minHeight: widget.minSize!,
-                  ),
+              constraints: BoxConstraints(
+                minWidth: widget.minSize ?? kCupertinoButtonMinSize[widget.sizeStyle]!,
```
I wonder if it'd be a good idea to change get these values from a static function using a switch statement, rather than a map, so that we don't have to use the `!` symbol?.

MitchellGoodwin reviewed (commented):
Thank you for the update! Left some comments.

kerberjg committed:
Adjusted default icon size in CupertinoButton

kerberjg committed:
Fixed CupertinoButton docs formatting

kerberjg committed:
Added tests for CupertinoButton styling constants

MitchellGoodwin commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -464,6 +524,14 @@ void main() {
     expect(decoration.color, isSameColorAs(CupertinoColors.systemBlue.darkColor));
   });
 
+  testWidgets('All CupertinoButton const maps keys\' match the available style sizes', (WidgetTester tester) async {
```
Good idea checking for a regression here

MitchellGoodwin replied:
The analyzer looks like it's complaining about the `/'`

MitchellGoodwin commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -10,20 +10,86 @@ import 'package:flutter/widgets.dart';
-const double _kCupertinoFocusColorSaturation = 0.835;
+// Measured against iOS (17) [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS).
+
+/// The size of a [CupertinoButton].
+/// Based on the iOS (17) [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS).
```
```suggestion
/// The size of a [CupertinoButton].
///
/// Based on the iOS (17) [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/buttons#iOS-iPadOS).
```
Dart doc comments need a line between the first sentence and the rest. The first line is a short summary that shows up as a preview when the detailed view isn't needed. It makes it easier to read in the code when there's a gap. This needs to happen throughout this PR.

MitchellGoodwin commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -464,6 +524,14 @@ void main() {
   });
 
+  testWidgets('All CupertinoButton const maps keys\' match the available style sizes', (WidgetTester tester) async {
+    for(final CupertinoButtonSize size in CupertinoButtonSize.values) {
```
```suggestion
    for (final CupertinoButtonSize size in CupertinoButtonSize.values) {
```
Analyzer is complaining about this

MitchellGoodwin reviewed (commented):
Left some nits. This is getting close. Sorry if this is a slog, there's a lot of process around public style changes like this.

kerberjg committed:
Moved CupertinoButton constants to constants.dart

kerberjg committed:
Fixed formatting

kerberjg committed:
Updated CupertinoButton disabled state colors

kerberjg commented:
@MitchellGoodwin no worries! It's a good opportunity to learn, now I know what to watch out for next time :)

I rebased on main and pushed the latest changes based on your feedback.

Earlier the CI reported a mismatch with the goldens due to the default icon size change, shall I go ahead and just update them?

MitchellGoodwin commented:
> Earlier the CI reported a mismatch with the goldens due to the default icon size change, shall I go ahead and just update them?

I think you need to be a member of flutter-hackers to be able to approve golden file differences, so one of us will approve them. Typically I like to wait until the PR is in a done state before I approve them. When I looked at them before the differences were all as expected.

victorsanni commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -442,6 +480,28 @@ void main() {
     );
     expect(textStyle.color, isSameColorAs(CupertinoColors.systemBlue.darkColor));
 
+
```
Nit: extra newline here should be removed.

victorsanni commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -612,32 +680,57 @@ void main() {
     );
+    expect(actualIconTheme?.size, defaultIconTheme.size);
+
+    // small size
```
Nit: Comments should start with a capital letter and end with a period, as per the style guide.

victorsanni commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -612,32 +680,57 @@ void main() {
 
     IconThemeData? actualIconTheme;
 
+    // large size
```
Nit: Comments should start with a capital letter and end with a period, as per the style guide.

victorsanni reviewed (commented):
A couple of nits, but this is close to landing. Thanks once again for the patience.

kerberjg committed:
Fixed CupertinoButton tests comment style

kerberjg commented:
Nits addressed

<sup>*Deploy Fridays, let's go!*</sup>

victorsanni commented on packages/flutter/lib/src/cupertino/text_theme.dart:
```diff
@@ -39,6 +39,20 @@ const TextStyle _kDefaultActionTextStyle = TextStyle(
+// color here, as their implementation depends on the default value of the color
+// field.
+//
+// Values derived from https://developer.apple.com/design/resources/.
```
Is there a specific figma file these values were derived from? If there is, I'd recommend to link it directly.

victorsanni commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -129,17 +129,37 @@ void main() {
+    );
+    await tester.pumpWidget(boilerplate(child: const CupertinoButton(
+      onPressed: null,
+      // sizeStyle: CupertinoButtonSize.large,
```
Should this be removed? We typically don't merge commented out code.

kerberjg committed:
Added CupertinoTextTheme source links

Figma file for:
- `_kDefaultActionTextStyle`
- `_kDefaultActionSmallTextStyle`

MitchellGoodwin reviewed (approved):
LGTM as well! Thank you very much!

MitchellGoodwin commented:
Golden file differences were as expected from the different default styling.

kerberjg commented:
Thank you too! It was good practice to get this one merged, on to the next one! 😊
